docs(framework) Update how-to guides for TLS and node authentication (#4809)
adap/flower: framework/docs/source/how-to-authenticate-supernodes.rst

Flower has built-in support for authenticated SuperNodes that you can use to verify the
identities of each SuperNode connecting to a SuperLink. For increased security, node
authentication can only be used when encrypted connections (SSL/TLS) are enabled. Flower
node authentication works similar to how GitHub SSH authentication works:

- SuperLink (server) stores a list of public keys of known SuperNodes (clients)
- Using ECDH, both SuperNode and SuperLink independently derive a shared secret
- Shared secret is used to compute the HMAC value of the message sent from SuperNode to
  SuperLink as a token
- SuperLink verifies the token

.. note::

    This guide builds on the Flower App setup presented in the
    :doc:`how-to-enable-tls-connections` guide and extends it to introduce node
    authentication to the SuperLink ↔ SuperNode connection.

.. tip::

    Checkout the `Flower Authentication
    <https://github.com/adap/flower/tree/main/examples/flower-authentication>`_ example
    for a complete self-contained example on how to setup TLS and node authentication.

.. note::

    This guide covers a preview feature that might change in future versions of Flower.

> Generating authentication keys
> ------------------------------

Reviewer: ```suggestion
Generate authentication keys
----------------------------
```

in the example linked at the top of this guide.

.. code-block:: bash

    # In the example directory, generate the public/private key pairs
    $ ./generate_auth_keys.sh

This will generate the keys in a new ``keys/`` directory. By default it creates a key
pair for the SuperLink and one for each SuperNode. Copy this directory into the
directory of your app (e.g. a directory generated earlier via ``flwr new``).

Enable node authentication in SuperLink
---------------------------------------

To launch a SuperLink with SuperNode authentication enabled, you need to provide three
aditional files in addition to the certificates needed for the TLS connections. Recall
that the authentication feature can only be enabled in the presence of TLS.

.. code-block:: bash
    :emphasize-lines: 5,6,7

    flower-superlink \
        --ssl-ca-certfile certificates/ca.crt \
        --ssl-certfile certificates/server.pem \
        --ssl-keyfile certificates/server.key \
        --auth-list-public-keys keys/client_public_keys.csv \
        --auth-superlink-private-key keys/server_credentials \
        --auth-superlink-public-key keys/server_credentials.pub

> .. dropdown:: Understand the command

Reviewer 2: The dropdown is a cool idea but it doesn't show properly in Github (see screenshot below).

<img width="1007" alt="Screenshot 2025-01-14 at 23 06 15" src="https://github.com/user-attachments/assets/8c902d0e-d3ea-4f4c-a979-1281d4b4a1bf" />

=======================

Flower has built-in support for authenticated SuperNodes that you can use to verify the
identities of each SuperNode connecting to a SuperLink. For increased security, node
authentication can only be used when encrypted connections (SSL/TLS) are enabled. Flower
node authentication works similar to how GitHub SSH authentication works:

- SuperLink (server) stores a list of public keys of known SuperNodes (clients)
- Using ECDH, both SuperNode and SuperLink independently derive a shared secret
- Shared secret is used to compute the HMAC value of the message sent from SuperNode to
  SuperLink as a token
- SuperLink verifies the token

.. note::

    This guide builds on the Flower App setup presented in the
    :doc:`how-to-enable-tls-connections` guide and extends it to introduce node
    authentication to the SuperLink ↔ SuperNode connection.

.. tip::

    Checkout the `Flower Authentication
    <https://github.com/adap/flower/tree/main/examples/flower-authentication>`_ example
    for a complete self-contained example on how to setup TLS and node authentication.

>     Check out the :doc:`how-to-authenticate-supernodes` guide to learn more about adding

Reviewer 2: Why does the documentation need to refer to itself?

    an authentication layer to SuperLink ↔ SuperNode connections.

.. note::

    This guide covers a preview feature that might change in future versions of Flower.

Generating authentication keys
------------------------------

To establish an authentication mechanism by which only authorized SuperNodes can connect
to a running SuperLink, a set of key pairs for both SuperLink and SuperNodes need to be
created.

We have prepared a script that can be used to generate such set of keys. While using
these are fine for prototyping, we advice you to follow the standards set in your
team/organization and generated the keys and share them with the corresponding parties.
Refer to the **Generate public and private keys for SuperNode authentication** section

> in the example linked at the top of this guide.

Reviewer 2: I think we should add the example link here to make it easier for the user.

node authentication works similar to how GitHub SSH authentication works:

- SuperLink (server) stores a list of public keys of known SuperNodes (clients)
- Using ECDH, both SuperNode and SuperLink independently derive a shared secret
- Shared secret is used to compute the HMAC value of the message sent from SuperNode to
  SuperLink as a token
- SuperLink verifies the token

.. note::

    This guide builds on the Flower App setup presented in the
    :doc:`how-to-enable-tls-connections` guide and extends it to introduce node
    authentication to the SuperLink ↔ SuperNode connection.

.. tip::

    Checkout the `Flower Authentication
    <https://github.com/adap/flower/tree/main/examples/flower-authentication>`_ example
    for a complete self-contained example on how to setup TLS and node authentication.
    Check out the :doc:`how-to-authenticate-supernodes` guide to learn more about adding
    an authentication layer to SuperLink ↔ SuperNode connections.

.. note::

    This guide covers a preview feature that might change in future versions of Flower.

Generating authentication keys
------------------------------

> To sstablish an authentication mechanism by which only authorized SuperNodes can connect

Reviewer: ```suggestion
To establish an authentication mechanism by which only authorized SuperNodes can connect
```

        --ssl-keyfile certificates/server.key \
        --auth-list-public-keys keys/client_public_keys.csv \
        --auth-superlink-private-key keys/server_credentials \
        --auth-superlink-public-key keys/server_credentials.pub

.. dropdown:: Understand the command

    * ``--auth-list-public-keys``: Specify the path to a CSV file storing the public keys of all SuperNodes that should be allowed to connect with the SuperLink.
      | A valid CSV file storing known node public keys should list the keys in OpenSSH format, separated by commas and without any comments. Refer to the code sample, which contains a CSV file with two known node public keys.
    * | ``--auth-superlink-private-key``: the private key of the SuperLink.
    * | ``--auth-superlink-public-key``: the public key of the SuperLink.

.. note::

    Currently, there is no support for dynamically removing, editing, or adding known
    node public keys to a running SuperLink. To change the set of known nodes, you need
    to shut the SuperLink down, edit the CSV file, and start the SuperLink again.
    Support for dynamically changing the set of known nodes will be available in an

>     upcoming Flower release.

Reviewer: @danielnugraha this is not true any more, right? IIRC, SuperLink re-loads the list of known public keys

Conclusion
----------

You should now have learned how to start a long-running Flower SuperLink and SuperNode
with node authentication enabled. You should also know the significance of the private

> key and store it safely to minimize security risks.

Reviewer: ```suggestion
key and store it securely to minimize risks.
```
